src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    states.push(currentState);
+  });
+
+  return states;
```

Comment:
The variable name `currentState` is a bit misleading since it implies that it's a single state being modified over time. A better name would be `stateClone` or `stateCopy`, which indicates that you're working with a copy of the state at each step.